tests/server.py

```diff
@@ -599,6 +602,54 @@ def advance(self, amount: float) -> None:
             super().advance(0)
 
 
+def validate_connector(connector: tcp.Connector, expected_ip: str) -> None:
+    """Try to validate the obtained connector as it would happen when
+    synapse is running and the conection will be established.
+
+    This method will raise a useful exception when necessary, else it will
+    just do nothing.
+
+    This is in order to help catch quirks related to reactor.connectTCP,
+    since when called directly, the connector's destination will be of type
+    IPv4Address, with the hostname as the literal host that was given (which
+    could be an IPv6-only host or an IPv6 literal).
+
+    But when called from reactor.connectTCP *through* e.g. an Endpoint, the
+    connector's destination will contain the specific IP address with the
+    correct network stack class.
+
+    Note that testing code paths that use connectTCP directly should not be
+    affected by this check, unless they specifically add a test with a
+    matching HomeserverTestCase.lookups[HOSTNAME] = "IPv6Literal".
+    For an example of implementing such tests, see test/handlers/send_email.py.
+    """
+    destination = connector.getDestination()
+
+    def check_ip(
+        cls: Union[Type[ipaddress.IPv4Address], Type[ipaddress.IPv6Address]]
+    ) -> None:
+        """With this class we produce a more informative error if needed"""
+        try:
+            cls(expected_ip)
+        except Exception:
+            raise ValueError(
+                "Invalid IP type and resolution, got %s, expected: %s %s"
+                % (destination, expected_ip, cls)
+            )
+
+    # We use address.IPv{4,6}Address to check what the reactor thinks it is
+    # is sending but check for validity with with ipaddress.IPv{4,6}Address
+    # because they fail with IPs on the wrong network stack.
+    if isinstance(destination, address.IPv4Address):
+        check_ip(ipaddress.IPv4Address)
+    elif isinstance(destination, address.IPv6Address):
+        check_ip(ipaddress.IPv6Address)
```

Comment:
I think `check_ip` isn't doing too much here:

```suggestion
    if isinstance(destination, (address.IPv4Address, address.IPv6Address)):
        try:
            destination.__class__(expected_ip)
        except Exception:
            raise ValueError(
                "Invalid IP type and resolution, got %s, expected: %s %s"
                % (destination, expected_ip, destination.__class__.__name__)
            )
```

Reply:
I think I got the gist of what you are hinting at, let me know if the changes I'm about to commit address your concerns.

However, do note the comment (and let me know if it can be made clearer):
```
    # We use address.IPv{4,6}Address to check what the reactor thinks it is
    # is sending but check for validity with with ipaddress.IPv{4,6}Address
    # because they fail with IPs on the wrong network stack.
```

Twisted's `address.IPv{4,6}Address` classes do not actually perform any kind of check, hence why the check goes explicitly against `isinstance(destination, address.IPv4Address` but the validation goes against `ipaddress.IPv4Address`.

Here is a quick interactive shell sample of what I mean:
```
>>> from twisted.internet import address
>>> address.IPv4Address("TCP", host="::1", port=1234)
IPv4Address(type='TCP', host='::1', port=1234)
>>> address.IPv6Address("TCP", host="127.0.0.1", port=1234)
IPv6Address(type='TCP', host='127.0.0.1', port=1234, flowInfo=0, scopeID=0)
>>>
>>> import ipaddress
>>> ipaddress.IPv4Address("::1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/ipaddress.py", line 1304, in __init__
    self._ip = self._ip_int_from_string(addr_str)
  File "/usr/local/lib/python3.9/ipaddress.py", line 1191, in _ip_int_from_string
    raise AddressValueError("Expected 4 octets in %r" % ip_str)
ipaddress.AddressValueError: Expected 4 octets in '::1'
>>> ipaddress.IPv6Address("127.0.0.1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/ipaddress.py", line 1915, in __init__
    self._ip = self._ip_int_from_string(addr_str)
  File "/usr/local/lib/python3.9/ipaddress.py", line 1628, in _ip_int_from_string
    raise AddressValueError(msg)
ipaddress.AddressValueError: At least 3 parts expected in '127.0.0.1'
```

Note that the errors from ipaddress are super terrible, and that's why we raise a different exception.

By forcing the exception to happen in my validations I see following instead:

```
  File "tests/server.py", line 645, in validate_connector
    raise ValueError(
builtins.ValueError: Invalid IP type and resolution for IPv4Address(type='TCP', host='localhost', port=465). Expected ::1 to be IPv4Address
```

I think that would be at least somewhat helpful if a developer hits this failure in the test suite, particularly since that would bring them to `validate_connector`, whose comment along with this error message would bring them along the right path.

Reply:
The difference between `address.IPv4Address` and `ipaddress.IPv4Address` was completely lost to me -- this needs to be made much more obvious.

Reply:
This still holds with the modified function? (GH is showing the old code here)

```
    # We use address.IPv{4,6}Address to check what the reactor thinks it is
    # is sending but check for validity with ipaddress.IPv{4,6}Address
    # because they fail with IPs on the wrong network stack.
    cls_mapping = {
        address.IPv4Address: ipaddress.IPv4Address,
        address.IPv6Address: ipaddress.IPv6Address,
    }

    cls = cls_mapping.get(destination.__class__)
```